### PR TITLE
feat: adding ads component

### LINKED ISSUE #3315: [Feature] enable maxLength option for KEYVALUE_ARRAY type
## Summary
- Enable a maxLength option for KEYVALUE_ARRAY which sets an upper limit to the length of the array element. 
- If the maxLength is set to 1, then don't display the "+" button to add more elements.

## Motivation
- Requirement came up for dynamodb datasource form where only one URL and port were supported i.e maxLength=1 case.

### LINKED ISSUE #3779: [Feature] Need a framework to add sub text to label via editor.json / form.json
## Summary
Need support in editor.json / form.json to add sub-text like here (the Hint text): 
Screenshot 2021-03-30 at 7 16 18 PM

It can be used to provide useful info about how to use a query / datasource form field.

### LINKED ISSUE #9223: [Feature]: Provide support to show a small question mark icon next to any label in UQI forms.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

A question mark icon would be useful in case we want to convey some extra information to the user about that the concerned input field does. When the user hovers over the question mark a small text with info could be displayed. 

### Why should this be worked on?

Improve user experience.

### PR DESCRIPTION
**Pull request by rashmi-rai**
## Description

Fixes #9223
Fixes #3315 
Fixes #3779

#### - Added ads components in folllowing FormControl components:
- InputTextControl
- DropdownControl
- CheckboxControl
- SwitchControl
- KeyValueArrayControl

- FieldArrayControl already uses ads.

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Tested manually.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feat/uqi-ads-component 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.96 **(-0.02)** | 36.53 **(-0.43)** | 34.21 **(0.04)** | 55.5 **(-0.01)**
 :red_circle: | app/client/src/components/ads/Checkbox.tsx | 26.83 **(0)** | 13.64 **(-1.36)** | 0 **(0)** | 30.56 **(0)**
 :red_circle: | app/client/src/components/ads/Dropdown.tsx | 59.61 **(-5.27)** | 47.35 **(-3.49)** | 37.31 **(-3.37)** | 59.36 **(-5.35)**
 :green_circle: | app/client/src/components/ads/TextInput.tsx | 75.41 **(0)** | 48.67 **(0.07)** | 84.21 **(0)** | 74.55 **(0)**
 :sparkles: :new: | **app/client/src/components/editorComponents/form/fields/StyledFormComponents.tsx** | **68** | **15.91** | **0** | **68**
 :red_circle: | app/client/src/components/formControls/CheckboxControl.tsx | 70.59 **(5.88)** | 9.52 **(-90.48)** | 20 **(-13.33)** | 68.75 **(-8.17)**
 :green_circle: | app/client/src/components/formControls/DropDownControl.tsx | 54.17 **(14.64)** | 2.13 **(-12.68)** | 20 **(3.33)** | 52.17 **(11.14)**
 :red_circle: | app/client/src/components/formControls/DynamicInputTextControl.tsx | 86.54 **(-7.34)** | 39.29 **(-27.38)** | 83.33 **(0)** | 86.27 **(-7.48)**
 :red_circle: | app/client/src/components/formControls/DynamicTextFieldControl.tsx | 98.04 **(-0.07)** | 70.59 **(0)** | 88.89 **(0)** | 97.92 **(-0.08)**
 :red_circle: | app/client/src/components/formControls/FieldArrayControl.tsx | 48 **(-2)** | 61.54 **(0)** | 0 **(0)** | 48 **(-2)**
 :red_circle: | app/client/src/components/formControls/FilePickerControl.tsx | 43.14 **(-4.13)** | 14.29 **(-25.71)** | 8.33 **(-8.34)** | 46.67 **(-2.31)**
 :green_circle: | app/client/src/components/formControls/FixedKeyInputControl.tsx | 44.44 **(1.11)** | 28.57 **(6.35)** | 16.67 **(0)** | 50 **(8.62)**
 :red_circle: | app/client/src/components/formControls/InputTextControl.tsx | 40 **(-3.59)** | 11.11 **(0)** | 14.29 **(-2.38)** | 38.46 **(-3.65)**
 :red_circle: | app/client/src/components/formControls/KeyValueArrayControl.tsx | 29.23 **(-5.77)** | 3.36 **(-3.78)** | 6.67 **(-1.02)** | 29.51 **(-5.58)**
 :green_circle: | app/client/src/components/formControls/KeyValueInputControl.tsx | 57.69 **(-1.57)** | 20 **(3.33)** | 12.5 **(0)** | 58.33 **(-1.67)**
 :green_circle: | app/client/src/components/formControls/SwitchControl.tsx | 54.29 **(3.07)** | 50 **(7.14)** | 28.57 **(0)** | 64 **(-0.29)**
 :red_circle: | app/client/src/components/formControls/WhereClauseControl.tsx | 34.43 **(-0.49)** | 40 **(0)** | 0 **(0)** | 34.43 **(-0.49)**
 :red_circle: | app/client/src/icons/FormIcons.tsx | 57.89 **(-0.93)** | 100 **(0)** | 11.11 **(-1.39)** | 57.89 **(-0.93)**
 :red_circle: | app/client/src/pages/Editor/FormControl.tsx | 23.68 **(-26.32)** | 0 **(0)** | 0 **(0)** | 27.27 **(-27.28)**
 :red_circle: | app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx | 25.98 **(-0.28)** | 15.38 **(0)** | 2.78 **(-0.25)** | 28.89 **(-0.42)**
 :green_circle: | app/client/src/selectors/commentsSelectors.ts | 85.25 **(1.64)** | 64.71 **(2.95)** | 73.33 **(0)** | 90.59 **(2.35)**
 :red_circle: | app/client/src/utils/FormControlRegistry.tsx | 40.43 **(-1.24)** | 100 **(0)** | 6.25 **(0)** | 40.43 **(-1.24)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :x: | ~~app/client/src/components/editorComponents/form/fields/CheckboxField.tsx~~ | ~~55.56~~ | ~~0~~ | ~~0~~ | ~~55.56~~
 :x: | ~~app/client/src/components/editorComponents/form/fields/NumberField.tsx~~ | ~~61.54~~ | ~~0~~ | ~~20~~ | ~~58.33~~
 :x: | ~~app/client/src/components/formControls/InputNumberControl.tsx~~ | ~~53.85~~ | ~~100~~ | ~~16.67~~ | ~~52~~
 :x: | ~~app/client/src/components/formControls/StyledControls.tsx~~ | ~~39.29~~ | ~~82.35~~ | ~~0~~ | ~~39.29~~</details>

### CODE REVIEW

File: app/client/src/components/editorComponents/form/fields/StyledFormComponents.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import React from "react";
+import styled from "styled-components";
+import { Colors } from "constants/Colors";
+import { ControlProps } from "components/formControls/BaseControl";
+
+//Styled help text, intended to be used with Form Fields
+export const StyledFormInfo = styled.span<{ config?: ControlProps }>`
```
Comment (rashmi-rai) [inline app/client/src/components/editorComponents/form/fields/StyledFormComponents.tsx:7]: add comments for checks

File: app/client/src/components/formControls/BaseControl.tsx [hunk 2]
```diff
@@ -41,18 +41,29 @@ export interface ControlProps extends ControlData, ControlFunctions {
...
+  initialValue?: string | boolean | number;
+  info?: string; //helper text
   isRequired?: boolean;
   conditionals: string;
   hidden?: HiddenType;
   placeholderText?: string;
   schema?: any;
+  errorText?: string;
+  showError?: boolean;
+  showTooltip?: boolean;
```
Comment (rashmi-rai) [inline app/client/src/components/formControls/BaseControl.tsx:62]: handle it via tooltipText

File: app/client/src/components/formControls/DropDownControl.tsx [hunk 3]
```diff
@@ -1,121 +1,76 @@
...
+    ) || {};
+  return (
+    <Dropdown
+      dontUsePortal={false}
+      dropdownMaxHeight="250px"
+      errorMsg={props.props?.errorText}
+      fillOptions
+      helperText={props.props?.info}
+      onSelect={props.input?.onChange}
+      {...props}
```
Comment (rashmi-rai) [inline app/client/src/components/formControls/DropDownControl.tsx:62]: test if {...props} is required

File: app/client/src/components/formControls/DynamicInputTextControl.tsx [hunk 4]
```diff
@@ -62,10 +61,7 @@ export function InputText(props: {
     customStyle = props.customStyles;
   }
   return (
-    <div style={customStyle}>
-      <FormLabel>
-        {label} {isRequired && "*"}
-      </FormLabel>
+    <div style={{ ...customStyle, width: "50vh", minHeight: "34px" }}>
```
Comment (rashmi-rai) [inline app/client/src/components/formControls/DynamicInputTextControl.tsx:64]: set width and minheight conditionally

File: app/client/src/components/formControls/FilePickerControl.tsx [hunk 5]
```diff
@@ -132,16 +131,9 @@ class FilePickerControl extends BaseControl<FilePickerControlProps> {
...
 
-    return (
-      <>
-        <FormLabel>
-          {label} {isRequired && "*"}
-        </FormLabel>
-        <Field component={FieldFileInput} name={configProperty} />
-      </>
-    );
+    return <Field component={FieldFileInput} name={configProperty} />;
```
Comment (rashmi-rai) [inline app/client/src/components/formControls/FilePickerControl.tsx:136]: replace filepicker by ads/filepicker

File: app/client/src/components/formControls/InputTextControl.tsx [hunk 6]
```diff
@@ -29,51 +32,51 @@ export function InputText(props: {
...
+  },
+) {
+  return (
+    <TextInput
+      dataType={props.dataType}
+      disabled={props.disabled || false}
+      onChange={props.input.onChange}
+      placeholder={props.placeholder}
+      value={props.input.value}
+      {...props.input}
```
Comment (rashmi-rai) [inline app/client/src/components/formControls/InputTextControl.tsx:68]: test without spread

File: app/client/src/components/formControls/KeyValueArrayControl.tsx [hunk 7]
```diff
@@ -58,30 +76,36 @@ function KeyValueRow(props: KeyValueArrayProps & WrappedFieldArrayProps) {
...
+        if (valueName && Array.isArray(valueName) && valueName?.length)
+          valueTextFieldName = `${field}.${valueName[1]}`;
+
         const otherProps: Record<string, any> = {};
         if (
           props.actionConfig &&
           props.actionConfig[index].description &&
           props.rightIcon
         ) {
+          //TODO: may not be in use
```
Comment (rashmi-rai) [inline app/client/src/components/formControls/KeyValueArrayControl.tsx:108]: remove rightIcon

File: app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx [hunk 8]
```diff
@@ -779,6 +796,93 @@ class DatasourceRestAPIEditor extends React.Component<Props> {
       </>
     );
   };
+
+  // All components in formControls must be rendered via FormControl.
+  // FormControl is the common wrapper for all formcontrol components and contains common elements i.e. label, subtitle, helpertext
+  renderInputTextControlViaFormControl(
```
Comment (rashmi-rai) [inline app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx:802]: Check with @albinAppsmith

File: app/client/src/pages/Editor/FormControl.tsx [hunk 9]
```diff
@@ -20,14 +30,133 @@ function FormControl(props: FormControlProps) {
...
+        {props.multipleConfig?.map((config) => {
+          return renderFormConfigTop({ config });
+        })}
+      </div>
+    );
+    bottom = props.multipleConfig?.map((config) => {
+      return renderFormConfigBottom({ config });
+    });
+    return (
+      <>
```
Comment (rashmi-rai) [inline app/client/src/pages/Editor/FormControl.tsx:68]: what is top, bottom and children?

File: app/client/src/pages/Editor/FormControl.tsx [hunk 10]
```diff
@@ -20,14 +30,133 @@ function FormControl(props: FormControlProps) {
...
+      </>
+    );
+  }
+
+  return (
+    <div>
+      <div
+        style={{
+          // TODO: replace condition with props.config.dataType === "TOGGLE"
+          display:
```
Comment (rashmi-rai) [inline app/client/src/pages/Editor/FormControl.tsx:81]: add comments for controlType check

File: app/client/src/utils/FormControlRegistry.tsx [hunk 11]
```diff
@@ -63,17 +63,23 @@ class FormControlRegistry {
...
-      buildPropertyControl(controlProps: KeyValueArrayProps): JSX.Element {
+      buildPropertyControl(
+        controlProps: KeyValueArrayControlProps,
+      ): JSX.Element {
         return <KeyValueArrayControl {...controlProps} />;
       },
     });
     FormControlFactory.registerControlBuilder("FILE_PICKER", {
       buildPropertyControl(controlProps: FilePickerControlProps): JSX.Element {
+        //TODO: may not be in use
```
Comment (rashmi-rai) [inline app/client/src/utils/FormControlRegistry.tsx:74]: filepicker is being used in dynamodb

File: app/client/src/components/formControls/KeyValueArrayControl.tsx (hunk 7 above)
Comment (rashmi-rai) [reply app/client/src/components/formControls/KeyValueArrayControl.tsx]: removed

**Review by ayushpahwa — DISMISSED** [review]

File: app/client/src/components/ads/Dropdown.tsx [hunk 12]
```diff
@@ -190,10 +231,11 @@ const Selected = styled.div<{
...
     background: ${(props) =>
-      props.hasError
-        ? Colors.FAIR_PINK
-        : props.theme.colors.dropdown.hovered.bg};
-  }
+      !props.isMultiSelect
+        ? props.hasError
+          ? Colors.FAIR_PINK
+          : props.theme.colors.dropdown.hovered.bg
+        : "white"}
```
Comment (nidhi-nair) [inline app/client/src/components/ads/Dropdown.tsx:238]: Nit: Shouldn't this be coming from constants as well?

File: app/client/src/pages/Editor/FormControl.tsx [hunk 13]
```diff
@@ -20,14 +30,135 @@ function FormControl(props: FormControlProps) {
...
+          <FormInfoText config={props.config}>{subtitle}</FormInfoText>
+        )}
+      </FormLabel>
+      {urlText && (
+        <FormInputAnchor href={url} target="_blank">
+          {urlText}
+        </FormInputAnchor>
+      )}
+      {displayType && (
+        <FormInputSwitchToJsonButton type="button">
```
Comment (nidhi-nair) [inline app/client/src/pages/Editor/FormControl.tsx:148]: Is this simply a plug for later implementation?

File: app/client/src/pages/Editor/FormControl.tsx [hunk 14]
```diff
@@ -20,14 +30,135 @@ function FormControl(props: FormControlProps) {
...
+
+  if (props.multipleConfig?.length) {
+    top = (
+      <div style={{ display: "flex" }}>
+        {props.multipleConfig?.map((config) => {
+          return renderFormConfigTop({ config });
+        })}
+      </div>
+    );
+    bottom = props.multipleConfig?.map((config) => {
```
Comment (nidhi-nair) [inline app/client/src/pages/Editor/FormControl.tsx:66]: What does multiple config signify here?

File: app/client/src/pages/Editor/FormControl.tsx (hunk 13 above)
Comment (rashmi-rai) [reply app/client/src/pages/Editor/FormControl.tsx:148]: yes. will add component for JSON and Form View later.

File: app/client/src/components/ads/Dropdown.tsx (hunk 12 above)
Comment (rashmi-rai) [reply app/client/src/components/ads/Dropdown.tsx:238]: There are form elements that are composed of more than one ads component.
For e.g. `KeyValueArrayControl` renders 2 `TextInput`, each requires a config hence it uses `multipleConfig` whereas the simpler ones like `InputTextControl` will uses `config` only.

File: app/client/src/pages/Editor/FormControl.tsx (hunk 14 above)
Comment (rashmi-rai) [reply app/client/src/pages/Editor/FormControl.tsx:66]: There are form elements that are composed of more than one ads component.
For e.g. `KeyValueArrayControl` renders 2 `TextInput`, each requires a config hence it uses `multipleConfig` whereas the simpler ones like `InputTextControl` will uses `config` only.

File: app/client/src/components/ads/Dropdown.tsx (hunk 12 above)
Comment (rashmi-rai) [reply app/client/src/components/ads/Dropdown.tsx]: added const here

**Comment by RakshaKShetty** [comment]
Hi @rashmi-rai following are the few issues found:

**Comment by rashmi-rai** [comment]
Hi @RakshaKShetty 
I have made the following fixes:
- [x] In Elastic search Db the user is not able to add Host URL details->Critical
- [x] When there is a error in the host address and then if the user add the correct Host the host name is displayed with red highlight

I will be fixing the following via validations PR:
- [ ] In Snowflake the Default schema should be a mandatory field 
- [ ] In S3 The Input Filed should be made mandatory as in Snowflake and save should be disabled until the mandatory fields are added(needs to be taken care of in all the Database)
- [ ] In redshift datasource-> When a user adds the file name is large the file picker input box seems to be broken 
Following may not be required.
- [ ] The user should not be allowed to enter characters in the Port input box. 

I will get it fixed by a different PR. Changes form back-end will be required.
- [ ] In API Pane under the setting tab the text under the check box should be aligned just below the above text

**Review by ayushpahwa — DISMISSED** [review]

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**